### PR TITLE
Fix issue 508

### LINKED ISSUE #508: Undesired handling of complex values when the expectation of an observable has zero real part
There are valid cases where the expectation value of some observable gives a real part that is zero. When this is the case, and there is some vanishingly small, but not exactly zero imaginary component the `safe_real` function applied within `expect` returns an undesirable error.

To any sensible user, the value of `safe_real(0 + 1e-323im)` when interpreted as an expectation value should evaluate to `0.0`. I'm not sure what the fix for this would be, but perhaps a more intelligent way of detecting values that are safe to convert to real needs to be considered.

To reproduce

```
julia> using YaoBlocks
julia> YaoBlocks.safe_real(0 + 1e-323im)
ERROR: Can not convert number 0.0 + 1.0e-323im to real due to its large imaginary part.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] safe_real(x::ComplexF64)
   @ YaoBlocks ~/.julia/packages/YaoBlocks/BlXhZ/src/utils.jl:319
 [3] top-level scope
   @ REPL[4]:1
```

### PR DESCRIPTION
**Pull request by GiggleLiu**
fix #508 by setting a looser threshold.

### CODE REVIEW

File: lib/YaoBlocks/src/utils.jl [hunk 1]
```diff
@@ -315,7 +315,7 @@ Base.vec(et::EntryTable) = Vector(et)
 # convert a (maybe complex) number x to real number.
 function safe_real(x)
     img = imag(x)
-    if !(iszero(img) || isapprox(x - im*img, x))
+    if !(iszero(img) || isapprox(x - im*img, x; atol=1e-15))
```
[the remark this answers is not in the document]
Comment (GiggleLiu) [reply lib/YaoBlocks/src/utils.jl:318]: Using eps might be too strict.
```julia
julia> eps(0.0)
5.0e-324
```

But you are right, eps may not be defined for e.g. symbolic types. Maybe implementing two interfaces is better.

[the remark this answers is not in the document]
Comment (GiggleLiu) [reply lib/YaoBlocks/src/utils.jl:318]: Oops, defining two functions cause `Dual` number input run into the other function. Got an error in the AD tests.

**Comment by Roger-luo** [comment]
I don't like this change, this is too heurestic, we should really just let user decide what precision they want. And only guarantee on throw or nothrow behaviour.

**Comment by jlbosse** [comment]
In what case would the proposed behaviour not be desirable? Grepping through the code `safe_real(x::T)` is currently only used in situations where we would expect `x` to be real up to precision `eps(T)` since it is the output of sum large-ish calculation that would be exactly real if we had no floating point error.

Its behaviour is i) return its input if `x` is real, ii) return the real part of `x` if the imaginary part is zero up to `eps(T)` if `eps(T)` is defined and iii) return the real part of `x` if the imaginary part is exactly zero if `eps(T)` is not defined. To me this is (with the exception of _maybe_ iii) ) exactly the behaviour I would find sensible. And iii) would fail loudly if we get something with tiny, but non-zero imaginary part.

**Comment by GiggleLiu** [comment]
The only issue is, the eps, which is defined by the gap between two float point numbers, might be a bit too strict. What about using `sqrt(eps(T))` or `10 * eps(T)`?
I am fine to make it a keyword argument so that users can tune it when the program errors unexpectedly.

**Comment by jlbosse** [comment]
I personally would be happy with either option, I guess do whichever is more standard when doing numerical computations?

**Comment by Roger-luo** [comment]
The problem is we cannot know how large this `eps` means fine. This is exactly why you have `rtol` and `atol` in `isapprox` to control such with a good default if you want an approximate behaviour. What I'm proposing is either:

a) forward these two options in `expect`;

```julia
expect(...; atol=..., rtol=...)
```

b) if user turns on `nothrow=True`, fallback to the unsafe `real` conversion and let users decide what they want to do with the result

```julia
expect(...; nothrow=True) # this is explicitly saying the calculation will not throw and may not be safe
```

**Comment by GiggleLiu** [comment]
Ok, I will assume users know what they are doing. I will throw a warning for large floating point errors to let user know they might want to use `sandwich` function instead.

@Roger-luo I think the interface that you proposed is too complicated. When the inputs are non-floating point types, it does not make sense to set eps.

**Comment by Roger-luo** [comment]
> I will throw a warning for large floating point errors to let user know they might want to use sandwich function instead. Then expect will never throw error.

that sounds good to me.

File: lib/YaoBlocks/src/utils.jl [hunk 2]
```diff
@@ -313,10 +313,11 @@ SparseArrays.sparse(et::EntryTable) = SparseVector(et)
 Base.vec(et::EntryTable) = Vector(et)
 
 # convert a (maybe complex) number x to real number.
-function safe_real(x)
+function safe_real(x::Complex{T}) where T
     img = imag(x)
-    if !(iszero(img) || isapprox(x - im*img, x))
-        error("Can not convert number $x to real due to its large imaginary part.")
+    if !iszero(img) && !(hasmethod(eps, Tuple{Type{T}}) && isapprox(x - im*img, x; rtol=10*eps(T), atol=10*eps(T)))
+        @warn "Fail to convert number $x to real due to the nonzero imaginary part: $img."
```
Comment (Roger-luo) [inline lib/YaoBlocks/src/utils.jl:319]: add a hint like "maybe you want to use sandwitch`?

Comment (GiggleLiu) [reply lib/YaoBlocks/src/utils.jl]: Good catch.

**Pull request merged**